kroxylicious-filters/kroxylicious-encryption/src/main/java/io/kroxylicious/filter/encryption/inband/InBandKeyManager.java

```diff
@@ -103,6 +112,11 @@ private Supplier<CompletionStage<KeyContext>> makeKeyContext(@NonNull K kekId) {
                             // or we need mutex
                             // or we externalize the state
                             AesGcmEncryptor.forEncrypt(new AesGcmIvGenerator(new SecureRandom()), dekPair.dek()));
+                }).toCompletableFuture()
+                // todo wire in a scheduler so we can delay/jitter DEK creation attempts
```

Comment:
Is that still a concern if the caffine cache has sorted things so that we only have a single future in flight no matter how many requests hit the cache?

Reply:
Backing off within the N dek creation attempts might still be a good idea.

Reply:
I don't think we can really rely on Caffeine to guarantee a single request is in flight. It only needs for a kek-per-topic pattern and we're going to in the situation of needing multiple concurrent KMS calls. That reminds me that the operations  we have right now in the `Kms` interface don't support batching, even though some KMSs support batching for some operations. But even with batching I think we should support some random exponential backoff to avoid a thundering herd problem for the KMS.

Reply:
I should probably have specified that this is per-KEK. Caffeine will take care of only having one async load in progress per key. So the current implementation would support multiple EDEKs being retrieved concurrently for different KEKs.

Reply:
Ok, I'm with you now. 

Caffine is ensuring we only have at most one pending future per cacheKey but we are also worried about the number of cacheKeys being used in parallel.

Reply:
Created issue #794 to think about implementing some higher-level rate limiting or other resilience measures. But that's something for after the initial work.